### PR TITLE
Fix reading `diagnostics` from undefined

### PR DESCRIPTION
**Pull request by ghengeveld**
Refactored the CLI, pulling context initialization up out of `runBuild` into `runAll` so that `ctx.options` is available outside of `runBuild`. Clarified how the exported function "entry points" are used and made `runBuild` "private" so that it must be tested _through_ `runAll` rather than directly.

All in all this is more than I bargained for since my initial goal was just to fix `ctx.options.diagnostics` (since `ctx.options` could be undefined before), but overall I think it's a good improvement, mostly for type safety.

Also upgraded https-proxy-agent because it wouldn't compile otherwise.

<details>
  <summary>📦 Published PR as canary version: <code>7.4.1--canary.840.6612535904.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install chromatic@7.4.1--canary.840.6612535904.0
  # or 
  yarn add chromatic@7.4.1--canary.840.6612535904.0
  ```
</details>

### CODE REVIEW

**Review by tmeasday — APPROVED** [review]
A good improvement for sure! I still don't like how keys are added to the context at random points later (e.g. `ctx.git`) but this is a big step forward!

File: node-src/index.ts [hunk 1]
```diff
@@ -102,11 +131,42 @@ export async function run({
...
+    return;
+  }
+
+  if (!isContext(ctx)) return;
+
   // Run these in parallel; neither should ever reject
   await Promise.all([runBuild(ctx), checkForUpdates(ctx)]);
 
-  if (ctx.exitCode === 0 || ctx.exitCode === 1) {
+  if ([0, 1].includes(ctx.exitCode) && 'options' in ctx) {
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply node-src/index.ts:169]: Ah yes, leftover from an earlier iteration.

File: node-src/index.ts [hunk 2]
```diff
@@ -102,11 +131,42 @@ export async function run({
...
+    ctx.log.error(fatalError(ctx, [e]));
+    ctx.extraOptions?.experimental_onTaskError?.(ctx, {
+      formattedError: fatalError(ctx, [e]),
+      originalError: e,
+    });
+    setExitCode(ctx, exitCodes.INVALID_OPTIONS, true);
+    return;
+  }
+
+  if (!isContext(ctx)) return;
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply node-src/index.ts:164]: Yes it should, though technically should never happen to begin with (it's just a type guard).

File: node-src/index.ts [hunk 3]
```diff
@@ -37,12 +53,32 @@ interface Output {
...
+>;
+
+const isContext = (ctx: InitialContext): ctx is Context => 'options' in ctx;
+
+// Entry point for the CLI, GitHub Action, and Node API
 export async function run({
   argv = [],
   flags,
-  options,
+  options: extraOptions,
```
[the remark this answers is not in the document]
Comment (ghengeveld) [reply node-src/index.ts:81]: Just `Configuration` (from Zod schema) I suppose.

[the remark this answers is not in the document]
Comment (ghengeveld) [reply node-src/index.ts:81]: There's no such thing.

File: node-src/index.test.ts [hunk 4]
```diff
@@ -12,7 +12,7 @@ import parseArgs from './lib/parseArgs';
 import TestLogger from './lib/testLogger';
 import uploadFiles from './lib/uploadFiles';
 import { runAll } from '.';
-import { runBuild } from './runBuild';
+// import { runAll } from './runAll';
```
Comment (ghengeveld) [inline node-src/index.test.ts:15]: This needs to go

**Pull request merged**